### ISSUE TITLE
EnumerableKey support ?

### ISSUE THREAD

**Issue by blemasle** [agileobjects/AgileMapper#162]
Hi,

Is there any way to use the key of an enumerable during the mapping. In the same manner that `EnumerableIndex` can be used when mapping index based collection ?

I looked in the docs and here in issues but couldn't find anything. Sorry if I missed it :/

```csharp
class A { }

class B
{
	public string Something { get; set; }
}
	
Mapper
	.WhenMapping
	.From<A>()
	.To<B>()
	.Map(ctx => ctx.EnumerableIndex.ToString()).To(d => d.Something);
	//.Map(ctx => ctx.EnumerableKey.ToLower()).To(d => d.Something);

var result = Mapper.Map(new List<A> { new A() }).ToANew<List<B>>();
var result2 = Mapper.Map(new Dictionary<string, A>
{
	["Key"] = new A()
}).ToANew<Dictionary<string, B>>();
```

**Comment by SteveWilkes** [comment]
Hi!

You didn't miss anything - there's no straightforward way to do that out of the box - seems like a sensible thing to include though, so I'll look into adding it. Thanks for the suggestion!

In the meantime, you can do it like this:

```csharp
var result = Mapper
    .Map(new Dictionary<string, A> { ["Key"] = new A() })
    .ToANew<Dictionary<string, B>>(cfg => cfg
        .WhenMapping
        .From<A>()
        .To<B>()
        .Map(ctx => ctx
            .Parent
            .GetSource<Dictionary<string, A>>()
            .Keys
            .ElementAt(ctx.EnumerableIndex.Value))
        .To(t => t.Something));
```

...not the prettiest, but it works by reaching up into the parent context when `A` is being mapped to `B`, casting the source to an `Dictionary<string, A>`, and accessing the `EnumerableIndex` element of the source dictionary's `Keys` collection. I've used inline configuration as that configuration will error if `A` is being mapped to `B` outside of a dictionary.

Thanks again!

Steve

**Comment by blemasle** [comment]
Great, I'll keep an eye opened for that new feature then. In the meantime, your suggestion perfectly fits my use case, even if it's not the prettiest.

Thank you!

**Comment by SteveWilkes** [comment]
It's taken me a while, but this is now being implemented in [this branch](https://github.com/agileobjects/AgileMapper/tree/features/ElementIndex). Cheers!

**Comment by SteveWilkes** [comment]
This is now supported in the latest v1.6 branch code, release to follow. I've named the property `ElementKey`, and renamed `EnumerableIndex` to `ElementIndex`, marking `EnumerableIndex` as obsolete.

The configuration would now look like:

```csharp
var result = Mapper
    .Map(new Dictionary<string, A> { ["Key"] = new A() })
    .ToANew<Dictionary<string, B>>(cfg => cfg
        .WhenMapping
        .From<A>().To<B>()
        .Map(ctx => ctx.ElementKey)
        .To(t => t.Something));
```

...or just:

```csharp
Mapper..WhenMapping
    .From<A>().To<B>()
    .Map(ctx => ctx.ElementKey)
    To(t => t.Something));
```

Cheers!

Steve

**Comment by SteveWilkes** [comment]
And v1.6-preview5 is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper)! Thanks again for the suggestion :)

**Comment by blemasle** [comment]
Thanks a lot!

**Comment by SteveWilkes** [comment]
This is included in v1.6, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper/1.6.0).

Thanks again for the feedback!
